Codemirror jlayouts cleanup

dgrammatiko opened the pull request:
#### Something odd was there

This just reformats the code on the JLayouts and properly using the API for inline JS and CSS
#### Testing

Make sure that codemirror is still working as expected

dgrammatiko committed:
jlayouts cleanup

dgrammatiko committed:
CS

wilsonge commented:
Travis please

wilsonge commented:
@okonomiyaki3000 can you review this please as the codemirror expert here :)

@dgt41 what is the result here if someone has overriden these JLayouts in a template with close to the existing code?

dgrammatiko committed:
Travis DocBlock

dgt41 commented:
@wilsonge this will be introduced in 3.5, so there is no way someone ever used that JLayout

wilsonge commented on plugins/editors/codemirror/codemirror.php:
```diff
@@ -87,16 +82,11 @@ public function onInit()
 		$dispatcher->trigger('onCodeMirrorAfterInit', array(&$this->params));
 
-		return '';
+		return;
```
You don't need to return at the end of a function - just leave this line out :)

okonomiyaki3000 replied:
I thought this function was meant to return a string but maybe not. Of course that doesn't explain why it was returning boolean in some cases until now... If the `true` (near the top of the function) was causing any kind of problem, nobody would likely have noticed as that only gets executed when there are two instances of the editor on the same page. I suppose I should test this theory...

okonomiyaki3000 replied:
Indeed, returning a boolean is a bad idea. However, returning nothing at all seems to be fine.

okonomiyaki3000 commented on plugins/editors/codemirror/layouts/editors/codemirror/init.php:
```diff
@@ -10,79 +10,65 @@
+$modPath         = json_encode(JUri::root(true) . '/' . $modePath . $extJS);
+
+JFactory::getDocument()->addScriptDeclaration(
+<<<JS
```
Personally, I hate heredoc and nowdoc. I don't know of any other place they're used in Joomla and I'd hate for this to set some kind of precedent. There's a bigger problem here though. Your IDE will see this big block of javascript as just some meaningless string. You won't get any syntax coloring or other useful features. The point of putting in a fake `<script>` tag was to trick editors into to displaying this code in a more useful way. You might call it a hack but it's a very useful one that does no harm.

dgt41 replied:
This is an actual rendering of PHPStorm, which one are you using?
![screen shot 2016-02-26 at 02 58 01](https://cloud.githubusercontent.com/assets/3889375/13339910/d20e09a0-dc34-11e5-9276-ee7db6eeb0db.png)

okonomiyaki3000 replied:
These days I use Atom, at the time I wrote that I was using SublimeText3. I'm sure there are very few editors that will make the kind of inference that PHPStorm is making here.

okonomiyaki3000 replied:
Does PHPStorm just look at a block of text and make some determination about what it is or does it have some rules about using specific `heredoc` identifiers (JS, CSS, etc) to denote the type of content?

dgt41 replied:
It should be js,css dependant, although I couldn't find anything on the customisation settings for that

dgt41 replied:
What I was trying to write was that if you put SCRIP, JAVASCRIPT or STYLESHEET the format falls back to simple text so JS and CSS should be special triggers in PHPStorm (testing suggests that, no documentation reference here).

But just to be fair here, the only part I really disliked was the return of the string from the JLayouts to just be included with the API, something that could be done in the layout themselves...

okonomiyaki3000 replied:
But that's what layouts do. They write output to a buffer that is then returned as a string. No matter what you change inside the layout file, a buffer will still be created and returned as a string. It might be empty or just white space and it might be discarded immediately but it still happens.

wilsonge replied:
I agree with @dgt41 - we have made the mistake in a few other places in the CMS of putting part of javascript in layouts. And this was (in retrospect) wrong. Because ideally what you want to do is to be able to treat it not like a helper file but as a standalone thing. In this case, this is responsible for injecting javascript and in the future you can add a html tag to show where the layout is called and finished (impossible at the moment because it inserts HTML comments in the middle of javascript - which is exactly what happens here)

Fedik replied:
I would move it to external js file :yum: ... (my 2 cents)

dgt41 replied:
@Fedik I wish that PR for data attributes @andrepereiradasilva made few weeks ago was available to use with external js. Still tho external js is the best solution here, will do that

Fedik replied:
do not stick the `data-` attributes (for the configuration) to the script tags, it could leads to another problems, there is an editor element ... just an advice :yum:

dgt41 replied:
article element? I lost you here

dgrammatiko committed:
CS

dgrammatiko committed:
CS

okonomiyaki3000 commented on plugins/editors/codemirror/codemirror.php:
```diff
@@ -153,8 +143,9 @@ public function onGetInsertMethod()
-		JFactory::getDocument()->addScriptDeclaration($js);
+		JFactory::getDocument()->addScriptDeclaration("
+		;function jInsertEditorText(text, editor) { Joomla.editors.instances[\"" . $id . "\"].replaceSelection(text); };
+		");
```
Don't do this. It will be a problem if you have a page with two CodeMirrors on it. 

There is a bigger problem if you have two **different** types of editors on the same page but that is a deeper problem in Joomla which will be much harder to fix. #5978

okonomiyaki3000 replied:
Also, in a case in which you need to create some js as a php string and (within that string) you need to insert a php string as a js string, you'd better do it like this:

```php
JFactory::getDocument()->addScriptDeclaration("
    ;function jInsertEditorText(text, editor) { Joomla.editors.instances[" . json_encode($id) . "].replaceSelection(text); };
 ");
```

If you just drop the string in between some "quotes", problems could occur. For example, if the inserted string also contained a `"`. Using `json_encode` will insure that everything is escaped and quoted properly. 

But anyway, don't do it here at all.

dgt41 replied:
The problem that exists in JEditor that cannot initiate more than one editors in a page I think is already mentioned few times but cannot be solved keeping B/C intact (at least that was my findings)...
About the placement of an untreated var in javascript: you are absolutely right, json_encode is the best way for this. Apparently this won't be the only place (tinyMCE and NONE are using unescaped vars, and that was not introduced by me ;) )

okonomiyaki3000 replied:
You **can** have multiple instances of the **same** editor on a page but that will not work the way you've done it here. You need to use the `editor` variable that is passed to the function by javascript not the `$id` that is passed by php. Using the js variable also eliminates the `json_encode` issue because it's a pure js solution.

Fedik replied:
here is I am fully agree with @okonomiyaki3000 
Please keep old `Joomla.editors.instances[editor]`
Because if you have multiple editors on the page the code `Joomla.editors.instances[\"" . $id . "\"]` will insert the text only for one of them.

okonomiyaki3000 commented on plugins/editors/codemirror/codemirror.php:
```diff
@@ -153,8 +213,9 @@ public function onGetInsertMethod()
-		$js = ";function jInsertEditorText(text, editor) { Joomla.editors.instances[editor].replaceSelection(text); }\n";
-		JFactory::getDocument()->addScriptDeclaration($js);
+		JFactory::getDocument()->addScriptDeclaration("
+		;function jInsertEditorText(text, editor) { Joomla.editors.instances[editor].replaceSelection(text); };
```
It's a minor thing but there shouldn't be a semicolon at the end of this javascript string.

dgt41 replied:
there is one `replaceSelection(text); }; `

okonomiyaki3000 replied:
Yes, I see it. It should not be there.

dgt41 replied:
ok done

okonomiyaki3000 replied:
I don't mean that in the sense of the stupid religious war about whether or not to allow js to do its automatic semicolon insertion. I mean that, following a named function declaration like this one, you do not use a semicolon. It obviously doesn't hurt anything but it has no meaning in the same way the the finally semicolon here has no meaning:

```javascript
for (var i = 0; i < length; i++) {
  // ...
};
```

okonomiyaki3000 commented on plugins/editors/codemirror/codemirror.php:
```diff
@@ -83,20 +95,66 @@ public function onInit()
+		.CodeMirror-activeline-background { background: " . $activeLineColor . "; }
+		.CodeMirror-matchingtag { background: " . $highlightMatchColor . "; }
+		"
+		);
```
This is exactly the kind of thing I wanted to avoid by using layouts. Now this is just a big string of css (which won't get any syntax coloring in most IDEs) sitting in the middle of a php class. To me, it's just gross. Also, you lose the ability to override it.

dgt41 replied:
Ok, I do understand the reason you did it the way you did, but jlayouts should be used just for html. I guess we need to find another way for javascript and css files.

dgrammatiko committed:
revert $id

dgt41 commented:
I reverted the last changes, so JLayouts are still used here, although at some point we should introduce a way to do javascript and css customisations with another approach, maybe JAssets, similar to JLayouts but for static files?

dgrammatiko committed:
remove semicolon

okonomiyaki3000 commented:
I see now that a problem with using layouts to generate js or css (or really to use them anywhere other than in a template or another layout file) is that they are able to generate some kind of debug message. That's really unfortunate and actually the way that debug message works is rather terrible, in my opinion. Obviously, we have to work with what we have. Personally, I'd still prefer to use layouts for all the benefits they offer and I can I propose a solution that still works when using the debug mode if anyone is interested.

dgt41 commented:
@okonomiyaki3000 if you can short out the debug problem then using JLayouts for js and css will be fine. It will be highly appreciated if you do a PR on that

okonomiyaki3000 commented:
I'll do it. But I can't promise the solution won't be slightly controversial. :wink:

dgt41 commented:
Even if it requires some extra stuff appended in the layout files (or anything similar), I think, will be acceptable as it will broaden the usage of JLayouts

okonomiyaki3000 committed:
Use output buffering around JLayoutHelper::render

`JLayoutHelper::render()` calls `JLayoutFile::render()` which (when in debug mode) will `echo` some unbuffered html. We obviously don't want that here so, catch it in an output buffer if need be and then simply disregard it.

okonomiyaki3000 commented:
@dgt41 I sent you a pull request with my quick solution to the issue. I'll think more about how the issue with layouts may be improved more generally and put in a different PR if I come up with anything good.

Commit:
Merge pull request #30 from okonomiyaki3000/patch-1

Thanks @okonomiyaki3000 Use output buffering around JLayoutHelper::render

dgrammatiko committed:
The editor element and the buttons should be plain html

dgrammatiko committed:
CS

wilsonge commented:
Where are we with this? Ideally I'd like to have this merged in asap

okonomiyaki3000 commented:
I'm OK with it. From what I can see there are two possible complaints but feel free to ignore them because you can never please everyone.

1) Doing the css and js as strings removes the ability of most editors to do syntax coloring which was part of the reason for putting them in layouts in the first place. (The more important reason was to make them overridable so as long as we have that, I'm fine.)

2) Layouts may be an inappropriate way to do js and css. I don't really see why as they provide an easy way for these things to be overridden and there is good reason to believe someone may want to override them. If this is truly and inappropriate use of layouts then it should be changed before there's any possibility of someone relying on them. Since I would really hate to lose the ability to override these things, maybe using JHtml could be a solution.

wilsonge commented:
I mean layouts aren't an ideal place to do this - but at the end of the day as you say it seems ideal and I think by ensuring that adding html comments won't break the css and js being returned means that we have a good compromise for putting CSS and JS in a standalone file. We're loosing the ability for the syntax coloring which kinda sucks - but the alternative of returning CSS and JS strings has even more pitfalls I fear :( I'll merge this tomorrow morning

okonomiyaki3000 commented:
On a side note, I really think something should be done about the debugging mode in JLayoutFile. Just echoing some code wrapped in a `<pre>` tag is not a good idea. Even if layouts were only ever used as they were apparently intended (from within template and layout files), this approach doesn't work all that well. For example, in a case that a layout file is rendered within another layout file (maybe in a loop to generate or list or table rows or something), wouldn't you end up with things showing up in a weird order? I think we need a better solution for this. Maybe use the debug console or something.

wilsonge commented:
I agree - although it appears that `<pre>` tag has existed for a hell of a long time without anyone coming across it... I mean we need *something* so that users know where to go to find the layout being used - that's the issue....

brianteeman commented:
Yeah I wasted a lot of time looking

okonomiyaki3000 commented:
I think a comment would work better than a `<pre>` for the reason that comments are legal anywhere, `<pre>` is not. For example, if your layout generated a table row and was obviously meant to be used within a `<table>`, a `<pre>` is going to be invalid. 

I also think it's not unreasonable to expect developers to pass some value in `$options` to specify what kind of output the layout is expected to generate. Then an appropriate comment syntax could be used. For even more control and future-proofing, allow them to pass in a format string for comments and just use that.

wilsonge commented:
We did try for 3.5 (well it was HTML comments rather than `/* foo */`) and all bootstrap modals immediately fell apart so we had to revert it :(

wilsonge commented:
@brianteeman Any reason why you reopened this?

brianteeman commented:
It wasnt closed on the issue tracker ;)

wilsonge commented:
Ah
